docs/source/torch.compiler_aot_inductor.rst

```diff
@@ -0,0 +1,184 @@
+
+Using AOTInductor, you can still author the model in Python. The following example demonstrates how to
+invoke ``aot_compile`` to transform the model into a shared library.
+
+.. note::
+
+   To execute the following code, it's essential to have a CUDA-enabled device on your machine.
+   If you do not possess a GPU, you can simply omit the ``.to(device="cuda")`` code within the snippet
+   below. In such a case, the script will compile the model code into a shared library that is optimized
+   for CPU execution.
+
+.. code-block:: python
+
+    import torch
+
+    class Model(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.fc1 = torch.nn.Linear(10, 16)
+            self.relu = torch.nn.ReLU()
+            self.fc2 = torch.nn.Linear(16, 1)
+            self.sigmoid = torch.nn.Sigmoid()
+
+        def forward(self, x):
+            x = self.fc1(x)
+            x = self.relu(x)
+            x = self.fc2(x)
+            x = self.sigmoid(x)
+            return x
+
+    with torch.no_grad():
+        m = Model().to("cuda")
+        example_inputs=(torch.randn(8, 10).to("cuda"),)
+        batch_dim = torch.export.Dim("batch", min=1, max=1024)
+        dynamic_shapes = {"x": {0: batch_dim}}
+        so_path = torch._export.aot_compile(m, example_inputs, dynamic_shapes=dynamic_shapes)
+
+    print(f"Compiled model into: {so_path}")
+    with open("model_so_path.txt", "w") as file:
+        file.write(so_path)
+
+In this illustrative example, the ``Dim`` parameter is employed to designate the first dimension of
+the input variable "x" as dynamic. Notably, the path and name of the compiled library remain unspecified,
+resulting in the shared library being stored in a temporay directory.
+To access this path from the C++ side, we save it to a file for later retrieval within the C++ code.
+For comprehensive details on the 'torch._export.aot_compile' API,
+you can refer to the code
+`here <https://github.com/pytorch/pytorch/blob/92cc52ab0e48a27d77becd37f1683fd442992120/torch/_export/__init__.py#L891-L900C9>`__.
+
+
+Inference in C++
+---------------------------
+
+Next, we use the following C++ file ``inference.cpp`` to load the shared library generated in the
+previous step, enabling us to conduct model predictions directly within a C++ environment.
+
+.. note::
+
+    Once more, it's imperative to have a CUDA-enabled device to execute the subsequent code.
+    In the absence of a GPU, it's necessary to make these adjustments in order to run it on a CPU:
+    1. Modify ``aoti_model_runner_cuda.h`` to ``aoti_model_runner.h``.
+    2. Change ``AOTIModelRunnerCuda`` to ``AOTIModelRunner``.
+    3. Eliminate ``.to(at::kCUDA)`` from the initialization of input tensors.
+
+.. code-block:: cpp
+
+    #include <fstream>
+    #include <iostream>
+    #include <vector>
+    #include <torch/torch.h>
+    #include <torch/csrc/inductor/aoti_model_runner_cuda.h>
+
+    int main() {
+        torch::NoGradGuard no_grad;
```

Comment:
You most likely want inference mode here and above.

Reply:
Pretty much all of the problems around inference mode had to do with the jit setting where we are in the middle of tracing / running our compile stack under inference mode.

But it should be 100% fine to turn inference mode here, where it will only be at runtime (and this is exactly the case where you'll see speedups from inference_mode over no_grad, if the generated inductor code here is using the pytorch dispatcher)

Reply:
I changed the C++ code to use `c10::InferenceMode`, but left the Python code to use `torch.no_grad()` just to be safe. @bdhirsh , do you happen to have a reproducible example for the problem we were discussing earlier? In the minimum, we should track the problem with an issue if not fixing it immediately.

Reply:
The last issue we hit was actually fixed: https://github.com/pytorch/pytorch/pull/112163/files#diff-48bd497885cd293ff4c37e382ec2f604b472adfe48d155da49db00ff99db54f2R771

The current state is basically

(1) Using inference_mode **during** tracing provides zero benefit of no_grad(), and is a bit fragile because of the way it interacts with the dispatcher. I think we've fixed all existing issues I know of, but I wouldn't be surprised if more pop up. We should just make this work in all cases in the long term though, and keep fixing bugs as they come up.

(2) Using inference_mode at runtime is strictly better than no_grad

(3) With `torch.compile`, we end up executing both the tracing and runtime all in one go, so people inevitably run the tracing bits of torch.compile under inference_mode

(4) with export though, there's a clean barrier between the user API for exporting the model and running the model. So it seems fine in the tutorial to use `no_grad()` during export - in the long term we'd like it to be functionally identical to exporting under `inference_mode()`, and in the medium term we know it's a bit less likely for users to run into issues